pkg/trace/stats/client_stats_aggregator.go

```diff
@@ -0,0 +1,283 @@
+	return []pb.ClientStatsPayload{trimHitsErrors(p)}
+}
+
+func (b *bucket) aggregateCounts(p pb.ClientStatsPayload) {
+	payloadAggKey := newPayloadAggregationKey(p.Env, p.Hostname, p.Version)
+	payloadAgg, ok := b.agg[payloadAggKey]
+	if !ok {
+		var size int
+		for _, s := range p.Stats {
+			size += len(s.Stats)
+		}
+		payloadAgg = make(map[bucketAggregationKey]*aggregatedCounts, size)
+		b.agg[payloadAggKey] = payloadAgg
+	}
+	for _, s := range p.Stats {
+		for _, sb := range s.Stats {
+			aggKey := newBucketAggregationKey(sb)
+			agg, ok := payloadAgg[aggKey]
+			if !ok {
+				agg = &aggregatedCounts{}
+				payloadAgg[aggKey] = agg
+			}
+			agg.hits += sb.Hits
+			agg.errors += sb.Errors
+			agg.duration += sb.Duration
+		}
+	}
+}
+
+func (b *bucket) flush() []pb.ClientStatsPayload {
+	if b.n == 1 {
+		return []pb.ClientStatsPayload{b.first}
+	}
+	return b.aggregationToPayloads()
+}
+
+func (b *bucket) aggregationToPayloads() []pb.ClientStatsPayload {
+	res := make([]pb.ClientStatsPayload, 0, len(b.agg))
+	for payloadKey, aggrCounts := range b.agg {
+		stats := make([]pb.ClientGroupedStats, 0, len(aggrCounts))
+		for aggrKey, counts := range aggrCounts {
+			stats = append(stats, pb.ClientGroupedStats{
+				Service:        aggrKey.service,
+				Name:           aggrKey.name,
+				Resource:       aggrKey.resource,
+				HTTPStatusCode: aggrKey.statusCode,
+				Type:           aggrKey.typ,
+				Hits:           counts.hits,
+				Errors:         counts.errors,
+				Duration:       counts.duration,
+			})
+		}
+		clientBuckets := []pb.ClientStatsBucket{
+			{
+				Start:    b.ts * 1e9, // to nanosecond
+				Duration: bucketDuration,
+				Stats:    stats,
+			}}
+		res = append(res, pb.ClientStatsPayload{
+			Hostname:         payloadKey.hostname,
+			Env:              payloadKey.env,
+			Version:          payloadKey.version,
+			Stats:            clientBuckets,
+			AgentAggregation: aggregatedCountPayload,
+		})
+	}
+	return res
+}
+
+type payloadAggregationKey struct {
+	env, hostname, version string
+}
+
+func newPayloadAggregationKey(env, hostname, version string) payloadAggregationKey {
+	return payloadAggregationKey{env: env, hostname: hostname, version: version}
+}
+
+type bucketAggregationKey struct {
+	service, name, resource, typ string
+	synthetics                   bool
+	statusCode                   uint32
+}
+
+func newBucketAggregationKey(b pb.ClientGroupedStats) bucketAggregationKey {
+	return bucketAggregationKey{
+		service:    b.Service,
+		name:       b.Name,
+		resource:   b.Resource,
+		typ:        b.Type,
+		synthetics: b.Synthetics,
+		statusCode: b.HTTPStatusCode,
+	}
+}
+
+func trimHitsErrors(p pb.ClientStatsPayload) pb.ClientStatsPayload {
+	for _, s := range p.Stats {
+		for i, b := range s.Stats {
+			b.Hits = 0
+			b.Errors = 0
+			b.Duration = 0
+			s.Stats[i] = b
+		}
+	}
+	return p
+}
+
+// aggregate separately hits, errors, duration
+// Distributions and TopLevelCount will stay on the initial payload
```

Comment:
Where is toplevel count used?

Reply:
It's used in TSA to define primaryName later in the pipeline